ls1intum/Artemis
# `Plagiarism checks`: Highlight relevant files when comparing programming exercise submissions

### Checklist
#### General


#### Client

### Motivation and Context
Closes #5066 

### Description
The file list as presented had two main issues:
- The files were listed in a random order, different for both students.
- The files for which actual plagiarism matches have been found were not highlighted.

Both issues make it hard to see which files need to be compared manually when checking for plagiarism. Therefore, this PR
- Highlights files with plagiarism matches in bold text.
- Sorts all files in a fixed order so that files with matches are grouped together at the top of the list. With the groups of files with/without matches the files are then sorted lexicographically.

Additional fixes:
- Use a proper dropdown element for the file selection.
- Limit the vertical size of the dropdown list and add scrolling so that it no longer gets cut off when there are many files. (E.g., in the before screenshot below there is actually at least one more file.)
- Closes #5141: Dropdown text not readable in dark mode, student names in sidebar hardly readable in dark mode

### Steps for Testing
Prerequisites:
- 1 Instructor
- 1 Programming exercise that has multiple student submissions so that a plagiarism report can be created.
- 1 Text exercise that has multiple student submissions so that a plagiarism report can be created.

1. Go to the plagiarism page for the programming exercise (white button ‘Check Plagiarism’ in the button group at the top of the exercise page).
2. Create a plagiarism report for this exercise.
    - Click on the ‘Plagiarism Detection’ header to open the options.
    - Tick ‘Generate JPlag Report’.
    - Generate the plagiarism report.
4. Select a student pair and check that the dropdown list of files is ordered as described above.
5. (optional) The JPlag-Report is a downloaded zip-File that contains a HTML document. In there you can select the same student pair and verify that all files listed there are highlighted in the dropdown in Artemis.
6. Open the plagiarism page for a text exercise. It should not have changed and still work without errors.

### Review Progress





#### Code Review
#### Manual Tests

### Test Coverage


| Class/File | Branch | Line |
|------------|-------:|-----:|
| `text-submission-viewer` | 50% | 75% |
| `split-pane-header` | 91% | 100% |

### Screenshots
Before:
![Screenshot 2022-05-14 at 06-12-36 Plagiarism Detection](https://user-images.githubusercontent.com/64250573/168413425-86f89cf2-9db1-484a-a7aa-8792bf80b137.png)

After:
The bold item in the list is just coincidentally the currently selected file, as in this case there is only one potential plagiarism case in this file. Multiple files would be marked bold otherwise.
![Screenshot 2022-05-23 at 14-52-44 Plagiarism Detection](https://user-images.githubusercontent.com/64250573/169825011-d56014d8-bf31-44a6-8f7f-5055e36364c1.png)
![Screenshot 2022-05-23 at 14-52-15 Plagiarism Detection](https://user-images.githubusercontent.com/64250573/169825007-32c6103a-fc8a-4ba6-be59-8374d842cd25.png)
